src/backend/kms/mod.rs

```diff
@@ -1010,49 +1061,58 @@ impl Device {
 
         Ok(output)
     }
-}
 
-pub fn source_node_for_surface(w: &WlSurface, dh: &DisplayHandle) -> Option<DrmNode> {
-    // Lets check the global drm-node the client got either through default-feedback or wl_drm
-    let client = dh.get_client(w.id()).ok()?;
-    if let Some(normal_client) = client.get_data::<ClientState>() {
-        return normal_client.drm_node.clone();
+    pub fn in_use(&self, primary: &DrmNode) -> bool {
+        &self.render_node == primary || !self.surfaces.is_empty() || !self.active_buffers.is_empty()
     }
-    // last but not least all xwayland-surfaces should also share a single node
-    if let Some(xwayland_client) = client.get_data::<XWaylandClientData>() {
-        return xwayland_client.user_data().get::<DrmNode>().cloned();
-    }
-    None
+}
+
+fn source_node_for_surface<'a>(
+    w: &WlSurface,
+    mut devices: impl Iterator<Item = &'a Device>,
+) -> Option<DrmNode> {
+    with_renderer_surface_state(w, |state| {
```

Comment:
I'm seeing a panic here in the `unwrap` within `with_renderer_surface_state` when I start an X client.

Reply:
https://github.com/Smithay/smithay/pull/1317